core/src/loader.rs

```diff
@@ -1604,38 +1609,116 @@ impl<'gc> Loader<'gc> {
+                Some(*loader_info.as_loader_info_object().unwrap())
+            } else {
+                None
+            };
+
+        if let Some(loader_info) = loader_info {
+            // Store the real movie into the `LoaderStream`, so that
+            // 'bytesTotal' starts returning the correct value
+            // (we previously had a fake empty SwfMovie).
+            // However, we still use `LoaderStream::NotYetLoaded`, since
+            // the actual MovieClip display object has not run its first
+            // frame yet.
+            loader_info.set_loader_stream(
+                LoaderStream::NotYetLoaded(movie.clone().unwrap(), Some(dobj.unwrap()), false),
+                uc.gc_context,
+            );
+        }
+
+        if let Some(dobj) = dobj {
+            if dobj.as_movie_clip().is_some() {
+                // We call these methods after we initialize the `LoaderInfo`, but before the
+                // add the the loaded clip as a child. The frame constructor should see
+                // 'this.parent == null' and 'this.stage == null'
+                dobj.post_instantiation(uc, None, Instantiator::Movie, false);
+                catchup_display_object_to_frame(uc, dobj);
+                // Movie clips created from ActionScript (including from a Loader) skip the next enterFrame,
+                // and consequently are observed to have their currentFrame lag one
+                // frame behind objects placed by the timeline (even if they were
+                // both placed in the same frame to begin with).
+                dobj.base_mut(uc.gc_context).set_skip_next_enter_frame(true);
+            }
+        }
+
+        if let Some(MovieLoaderEventHandler::Avm2LoaderInfo(loader_info)) = event_handler {
+            let domain = uc
+                .library
+                .library_for_movie(movie.unwrap())
+                .unwrap()
+                .avm2_domain();
+            let mut activation = Avm2Activation::from_domain(uc.reborrow(), domain);
+            let mut loader = loader_info
+                .get_public_property("loader", &mut activation)
+                .map_err(|e| Error::Avm2Error(e.to_string()))?
+                .as_object()
+                .unwrap()
+                .as_display_object()
+                .unwrap()
+                .as_container()
+                .unwrap();
+
+            // Note that we do *not* use the 'addChild' method here:
+            // Per the flash docs, our implementation always throws
+            // an 'unsupported' error. Also, the AVM2 side of our movie
+            // clip does not yet exist. Any children added inside the movie
+            // frame constructor will see an 'added' event immediately, and
+            // an 'addedToStage' event *after* the constructor finishes
+            // when we add the movie as a child of the loader.
+            loader.insert_at_index(&mut activation.context, dobj.unwrap(), 0);
```

Comment:
Is it the case that, whenever we have `Some(MovieLoaderEventHandler::Avm2LoaderInfo(loader_info))` and the other checks in this method, that we either should also have a `dobj` that is `Some`, or else we should panic? I am asking due to the `unwrap()` here, and that `dobj` seems like it might be `None` in other branches without panic'ing. It might well be the case that the changes here are good and relevant, I am asking due to me not understanding this part of the codebase.

Reply:
Yes, this will always be present here.